### PR TITLE
Docker provider

### PR DESCRIPTION
**Pull request by jefferai**
Please read this merge request message carefully; it is not expected that this be merged. This is being created for initial feedback.

Initial commit. This adds the initial bits of a Docker provider.
Docker's API is huge and only a small subset is currently implemented,
but this is expected to grow over time. This is enough to satisfy the
use cases of probably 95% of Docker users.

A few things are not yet tested, and are clearly marked with TODO. These
are very high on the priority list.

I'm preparing this initial pull request as a preview step for feedback.
My ideal scenario would be to develop this within a branch in the main
repository; the more eyes and testing and pitching in on the code, the
better (this would avoid a merge request-to-the-merge-request scenario,
as I figure this will be built up over the longer term, even before
a merge into master).

Unit tests do not exist yet. I really need
https://github.com/hashicorp/terraform/issues/950 and
https://github.com/hashicorp/terraform/issues/952 fixed before I can
properly instrument Docker at all, let alone trust the results of unit
tests. Working around this to the best of my abilities is also a reason
that each resource setting has a computed equivalent.

I also added a "make quickdev" option to skip the fetching step, which
speeds up builds significantly when you've just updated all of the
dependencies.

This code is copyright 2014 Akamai Technologies, Inc. <opensource@akamai.com>

### CODE REVIEW

**Comment by mitchellh** [comment]
Note: we're doing a bunch of merges for 0.3.7 right now and we won't get to this until 0.4.0 (which will start this week).

**Comment by jefferai** [comment]
@mitchellh Not a problem. While I'm fixing up the TODOs and bringing in the fixes you've been working on I've been rebasing. When you guys start taking a look I'll switch to merging if you wish.

File: builtin/providers/docker/resource_docker_container.go [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+				ForceNew: true,
+			},
+
+			// ForceNew is not true for image because we need to
+			// sane this against Docker image IDs, as each image
+			// can have multiple names/tags attached do it.
+			"image": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
```
Comment (phinze) [inline builtin/providers/docker/resource_docker_container.go]: Comment says "`ForceNew` is not true" but it is set. Which one is correct?

File: builtin/providers/docker/resource_docker_container_funcs.go [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+	return nil
+}
+
+func resourceDockerContainerUpdate(d *schema.ResourceData, meta interface{}) error {
+	if err := resourceDockerContainerDelete(d, meta); err != nil {
+		return fmt.Errorf("Error deleting container before recreation: %s", err)
+	}
+
+	return resourceDockerContainerCreate(d, meta)
+}
```
Comment (phinze) [inline builtin/providers/docker/resource_docker_container_funcs.go]: If you can only nuke-and-recreate in your Update, you may just want to omit it and set all your configurable attributes to `ForceNew` - that way your plans will be more clear about when they're going to `-/+`

File: scripts/build.sh [hunk 3]
```diff
@@ -19,9 +19,11 @@ GIT_DIRTY=$(test -n "`git status --porcelain`" && echo "+CHANGES" || true)
...
 XC_OS=${XC_OS:-linux darwin windows freebsd openbsd}
 
-# Install dependencies
-echo "==> Getting dependencies..."
-go get ./...
+# Install dependencies unless running in quick mode
+if [ "${TF_QUICKDEV}x" == "x" ]; then
+    echo "==> Getting dependencies..."
+    go get ./...
+fi
```
Comment (phinze) [inline scripts/build.sh]: Think you could peel this `QUICKDEV` feature out as a separate PR?

File: builtin/providers/docker/resource_docker_container.go (hunk 1 above)
Comment (jefferai) [inline builtin/providers/docker/resource_docker_container.go]: Oops. Will take it out.

File: builtin/providers/docker/resource_docker_container_funcs.go [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+		} else {
+			image = image + ":latest"
+		}
+	}
+
+	// The awesome, wonderful, splendiferous, sensical
+	// Docker API now lets you specify a HostConfig in
+	// CreateContainerOptions, but in my testing it still only
+	// actually applies HostConfig options set in StartContainer.
+	// How cool is that?
```
Comment (jefferai) [inline builtin/providers/docker/resource_docker_container_funcs.go]: Haha

File: scripts/build.sh (hunk 3 above)
Comment (jefferai) [inline scripts/build.sh]: Done... #1151

**Comment by phinze** [comment]
@jefferai seems like this might be nearly ready to merge, yeah? Just need to get some acceptance tests in there. :+1:

**Comment by mitchellh** [comment]
Starting to review and get this going for 0.4

File: builtin/providers/docker/config.go [hunk 5]
```diff
@@ -0,0 +1,24 @@
+package docker
+
+import dc "github.com/fsouza/go-dockerclient"
```
Comment (mitchellh) [inline builtin/providers/docker/config.go]: Nitpick I usually just do multiline imports from the get-go cause there is usually some future that requires more and it makes it easier to add.

File: builtin/providers/docker/resource_docker_container_funcs.go (hunk 4 above)
Comment (mitchellh) [inline builtin/providers/docker/resource_docker_container_funcs.go]: Haha, I'll remove the sarcasm from the final merge, but noted.

File: builtin/providers/docker/resource_docker_container_funcs.go [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceDockerContainerCreate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+
+	client, err := config.NewClient()
+	if err != nil {
+		return fmt.Errorf("Unable to connect to Docker: %s", err)
+	}
```
Comment (mitchellh) [inline builtin/providers/docker/resource_docker_container_funcs.go]: Any specific reason the client can't be initialized in the provider configuration function?

**Comment by mitchellh** [comment]
@jefferai This looks good, I can add the tests if you're averse to it. Is this ready to go? I noticed you haven't comitted in awhile but I think in IRC you mentioned you were doing more work on it.

**Comment by mitchellh** [comment]
Closing for #1329 which is just your commit + some test commits.

**Pull request closed without merging**